### PR TITLE
[Merged by Bors] - feat(CategoryTheory/Triangulated): t-structures

### PR DESCRIPTION
**Pull request by joelriou**
This PR defines t-structures on (pre)triangulated categories.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Triangulated/TStructure/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+use depending on the context.
+
+## TODO
+
+* define functors `t.truncLE n : C ⥤ C`,`t.truncGE n : C ⥤ C` and the
+associated distinguished triangles
+* promote these truncations to a (functorial) spectral object
+* define the heart of `t` and show it is an abelian category
+* define triangulated subcategories `t.plus`, `t.minus`, `t.bounded` and show
+that there are induced t-structures on these full subcategories
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Triangulated/TStructure/Basic.lean:34]:
```suggestion
* define functors `t.truncLE n : C ⥤ C`,`t.truncGE n : C ⥤ C` and the
  associated distinguished triangles
* promote these truncations to a (functorial) spectral object
* define the heart of `t` and show it is an abelian category
* define triangulated subcategories `t.plus`, `t.minus`, `t.bounded` and show
  that there are induced t-structures on these full subcategories
```

File: Mathlib/CategoryTheory/Triangulated/TStructure/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  have H_add : ∀ (a b c : ℕ) (_ : a + b = c) (_ : H a) (_ : H b), H c := by
+    intro a b c h ha hb n
+    rw [← h, Nat.cast_add, ← add_assoc]
+    exact (ha n).trans (hb (n+a))
+  intro a
+  induction' a with a ha
+  · exact H_zero
+  · exact H_add a 1 _ rfl ha H_one
+
+lemma GE_antitone : Antitone t.GE := by
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Triangulated/TStructure/Basic.lean:131]: Is it possible to obtain this proof from `LE_monotone` via the opposite category?

Comment (joelriou) [reply Mathlib/CategoryTheory/Triangulated/TStructure/Basic.lean:131]: Thanks for the suggestion! I wish I could, but the opposite `t`-structure is not so easy to define. `LE n` in the opposite category corresponds to `GE (-n)` in the original category, etc. Then, it seems that in order to show `LE_zero_le` and `GE_one_le` for the opposite category, we would basically need to know both `LE_monotone` and `GE_monotone` in a few cases for the original category...

Comment (jcommelin) [reply Mathlib/CategoryTheory/Triangulated/TStructure/Basic.lean:131]: Aah, good point.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**